### ISSUE TITLE
[HOLD for payment 2024-06-28] MEDIUM: Make auto-complete component work inline

### ISSUE THREAD

**Issue by puneetlath** [Expensify/App#16078]
Our existing auto-complete component works well, but because it has to live above the composer, it limits the size the composer can grow to when expanded. Ideally, the emoji component would work in-line, similar to Slack's so that the composer has the ability to grow to the full height. 

cc @perunt 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01dd591b8c1d02e181</li>
        <li>Upwork Job ID: 1664650244045303808</li>
        <li>Last Price Increase: 2023-06-02</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @puneetlath</details>

**Comment by perunt** [comment]
I'll take a look

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days.  eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by puneetlath** [comment]
@perunt how's it going with this one?

**Comment by puneetlath** [comment]
@perunt for the react native PR, is the plan to also try to get that merged upstream?

**Comment by perunt** [comment]
I'm not certain how the review process will unfold since it's usually a lengthy one, but I've submitted an upstream pull request

**Comment by puneetlath** [comment]
@perunt would you mind laying out what you see as the path to inline auto-completion? 

I see we have these two PRs, the first of which is merged, and the second of which is close to being able to merge. What's the plan after that? 

https://github.com/Expensify/react-native/pull/49
https://github.com/Expensify/App/pull/17331

**Comment by perunt** [comment]
@puneetlath Yes, the first PR has been merged, and the team at Callstack is currently updating the React Native version. These changes should be available in the app in the coming days and will bring those changes into the app. 
The second PR is currently in the review process. I've started working on implementing inline suggestions, but I need these changes to continue, so I've put it on hold. I think the block will be resolved as soon as we update the React Native version.

**Comment by puneetlath** [comment]
Ok cool!

**Comment by perunt** [comment]
To improve tracking of this task, I'll provide a live link to the PR that I'm waiting for [link](https://github.com/Expensify/App/pull/18507)

**Comment by puneetlath** [comment]
@perunt could you lay out what the next series of tasks will be now that the first PR has been merged? It would help us get a picture of the roadmap for getting from where we are to completion.

**Comment by perunt** [comment]
All the PRs I've been working on are done and merged. Some of them merged it into our latest React Native version. We're just waiting on one [last PR](https://github.com/Expensify/App/pull/18507), which bring that changes.
Next up, we're going to whip up a super handy component for inline suggestions in our text input. It'll be flexible, so we can also use it for stuff like mentions.
After we've got that component, we'll be using the same tool to suggest emojis. 
Hope it makes sense

**Comment by MonilBhavsar** [comment]
Hi, just a heads-up! Coming from https://github.com/Expensify/App/issues/15934. We have put that issue on hold for this, as we believe this issue will change behavior with suggestion modal

**Comment by puneetlath** [comment]
@0xmiroslav I went ahead and sent you a contract for the C+ review on the first PR.

**Comment by puneetlath** [comment]
Ok, paid @0xmiroslav for the first PR. 

@perunt when can we expect the next one? 😄

**Comment by perunt** [comment]
We're just waiting on [updating RN PR](https://github.com/Expensify/App/pull/18507)

**Comment by puneetlath** [comment]
Looks like there's a lot of active work happening in the RN PR.

**Comment by puneetlath** [comment]
@perunt you're still actively working on RN PRs for this, is that right?

**Comment by perunt** [comment]
Yes, since the Comment Linking in the review, I'm now fully committed to this task. As we've decided to abandon our own RN fork, it's going to take some time to integrate it into the core. Nonetheless, I can create an immediate solution by developing a patch package with the necessary changes. This would allow us to use it right away, and we can then proceed to integrate these changes upstream

**Comment by puneetlath** [comment]
That sounds great!

**Comment by puneetlath** [comment]
@perunt would you mind giving us a weekly update in this GH issue? So we can follow along with the progress.

**Comment by puneetlath** [comment]
Ok cool. Can you just link or copy that update here every week then. That way it's easy for @0xmiroslav and I to stay in the loop and the Overdue label won't get added. Thanks!

**Comment by perunt** [comment]
Still blocked by updating process to the 0.72 version and PRs, which stuck on the RN core team table

**Comment by puneetlath** [comment]
Looks like 0.72 is merged! https://github.com/Expensify/App/pull/18507

**Comment by puneetlath** [comment]
@perunt any movement?

**Comment by perunt** [comment]
still waiting for cursor[ position features](https://github.com/facebook/react-native/pull/36979#issuecomment-1620060057) to be merged

**Comment by puneetlath** [comment]
@perunt any updates on how it's going getting those updates merged?

**Comment by puneetlath** [comment]
@perunt any updates?

**Comment by puneetlath** [comment]
Still waiting on upstream feedback.

**Comment by puneetlath** [comment]
Still waiting on upstream feedback.

**Comment by puneetlath** [comment]
Still waiting on upstream feedback.

**Comment by puneetlath** [comment]
Still waiting on upstream feedback. @perunt can we expect any movement here or should we try another approach?

**Comment by puneetlath** [comment]
@perunt we raised a feature request, is that right?

**Comment by perunt** [comment]
Yes, we published [RFC](https://github.com/react-native-community/discussions-and-proposals/pull/730)

**Comment by puneetlath** [comment]
@perunt where do we go from here?

**Comment by quinthar** [comment]
Can we patch RN to make this work for us?

**Comment by puneetlath** [comment]
@perunt I believe you said in Slack that your plan is to patch RN for this, is that right? Can you provide an outline of your plan here?

**Comment by perunt** [comment]
Yes, you're right. David mentioned that he wants to see this feature ASAP, so we'll patch it and try to push this forward upstream. From next week, I'll do a patch And provide an update.

**Comment by puneetlath** [comment]
@perunt have you had a chance to start on that plan yet? If so, can you please share your proposed approach?

**Comment by perunt** [comment]
@puneetlath, I've landed on this. I reused old code for Android and tested it for any performance impact, and it looks okay. However, when I tried to apply the old code to iOS, it failed. It seems like they've updated the Fabric implementation of the text input, so the code needs to be adjusted. This also involves making changes in the upstream repository. But I think it makes sense to do it locally first, to test and release, and then push it upstream.
What do you think? Would you like me to fix it locally only?

**Comment by perunt** [comment]
@puneetlath I just pushed a patch for inline autosuggestion

**Comment by perunt** [comment]
@hoangzinh, we've discussed that with @puneetlath on Slack. He's good with this, and it's already implemented. As far as I know, he's out of the office until February 1st, so it makes sense to look into it before he returns. Merging this PR will enable me to deliver inline autosuggestions and mention features.

**Comment by allroundexperts** [comment]
Taking over the review!

**Comment by quinthar** [comment]
I'm coming into this one real late -- there is a huge thread.  Can someone describe the problem that this is trying to solve, and how we are going to solve it?

**Comment by quinthar** [comment]
I'm coming into this one real late -- there is a huge thread. @puneetlath can you please describe the problem that this is trying to solve, and how we are going to solve it?

**Comment by perunt** [comment]
I can explain what's happening now. Recently, we received `MarkdownTextInputProps` in main.
When I was working on auto-suggestions, I found some bugs ([link to thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1708595666868069?thread_ts=1707846163.324679&cid=C01GTK53T8Q)) that make the suggestions not work right on iOS. There's a PR for adding auto-suggestions ([link](https://github.com/Expensify/App/pull/36239)), but it needs the `MarkdownTextInputProps` to be updated first. If we merge it now, the suggestion box might blink or not show up on iOS, but it works on other platforms

**Comment by puneetlath** [comment]
Discussion here: https://expensify.slack.com/archives/C01GTK53T8Q/p1711638748757719

**Comment by puneetlath** [comment]
How's it going @perunt?

**Comment by puneetlath** [comment]
@perunt how is this going?

**Comment by mallenexpensify** [comment]
Any update here @perunt ? Are you able to prioritize this? Looks like we have [a handful of issues ](https://github.com/Expensify/App/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+%5BHOLD+%2316078%5D)held on this.  Thx

**Comment by perunt** [comment]
@mallenexpensify, the PR is in the review process. I don't know how long it could take. If you want, I can start fixing those issues with my changes right now.

**Comment by mallenexpensify** [comment]
> I can start fixing those issues with my changes right now.

@puneetlath would know better than I.  Thx

**Comment by puneetlath** [comment]
Woohoo! @perunt I think the next best one for you to work on would be adding the auto-suggestion component (for mentions and emojis) to the edit message dialog: https://github.com/Expensify/App/issues/19643

**Comment by alitoshmatov** [comment]
I was C+ reviewer on the PR. @puneetlath Can you assign me to handle payment here

**Comment by puneetlath** [comment]
Assigned! Are there any regression tests you think we should add for this?

**Comment by puneetlath** [comment]
It looks like @alitoshmatov ended up doing the review here. So I think @Ollyws isn't due anything, but let me know if I'm mistaken.

**Comment by alitoshmatov** [comment]
@puneetlath Yes this is new feature, we require a regression tests. Testing steps in PR looks good:

Testing Emoji Suggestions:

1. Open the chat interface.
2. Type an emoji trigger, starting with :sm followed by at least two more characters (e.g., :smile).
3. Verify that the suggestion box appears adjacent to the cursor.
4. Move the cursor to a new position in the text field.
5. Again, type an emoji-like text with at least two characters following the :sm prefix.
6. Confirm that the suggestion box appears next to the cursor as expected.
7. Testing Mention Suggestions:

Open the chat interface.
1. Initiate a mention by typing the @ symbol followed by any text (e.g., @username).
2. Ensure the suggestion box pops up next to the cursor.
3. Reposition the cursor within the text field.
4. Type another mention, starting with the @ symbol.
5. Check that the suggestion box is correctly displayed adjacent to the cursor.

**Comment by puneetlath** [comment]
Awesome, thanks. GH issue for regression test is here: https://github.com/Expensify/Expensify/issues/409359

Upwork offer is here: https://www.upwork.com/nx/wm/offer/102945166

Please ping me on this issue when you've accepted.

**Comment by alitoshmatov** [comment]
@puneetlath accepted the offer

**Comment by puneetlath** [comment]
Paid. Thanks everyone!
